### ISSUE TITLE
'if (max(abs(g)) < 1e-10) break' error after convergence issue and AICc change with control argument

### ISSUE THREAD

**Issue by tmerkling** [glmmTMB/glmmTMB#399]
Hi,

I'm trying to run binomial GLMMs with a binary response variable and 4 random intercepts and I have trouble fitting all the models.
Some models run fine (I'm using version 0.2.2.0 and R 3.4.4), some give a convergence warning `non-positive-definite Hessian matrix` that can get fixed by adding `control = glmmTMBControl(optCtrl = list(iter.max = 1000, eval.max = 1000), profile = TRUE)`,  but:
1)  others give an error message, after adding the control argument, that says:
 `Error in if (max(abs(g)) < 1e-10) break : missing value where TRUE/FALSE needed`
2) others still don't converge after adding the control argument
3) adding the control argument seems to influence the AICc (e.g. I've noticed a >1.5 AICc change!!)

Any idea what the first problem means? I think the lack of convergence comes from the fact that the mean probability of occurence of my response variable is quite low, is that possible?
Is the third problem known, and does it mean that the control argument should be used either for all or none of the models compared?
I've attached the data. 

This is one model that doesn't converge:
`mod_TMB_Qlty <-  glmmTMB(ProbDiv ~ stdQlty + stdLaying + (1|Year) + (1|Site) + (1|BandM) + (1|BandF), family = "binomial", control = glmmTMBControl(optCtrl = list(iter.max = 1000, eval.max = 1000), profile = TRUE), data =  IC_comp_data)`

I then tried to reduce the complexity by replacing the ID of the male and of the female by the Pair ID to have only 3 random intercepts, but then got the error message:
`mod_TMB_Qlty <-  glmmTMB(ProbDiv ~ stdQlty + stdLaying + (1|Year) + (1|Site) + (1|PairID), family = "binomial", control = glmmTMBControl(optCtrl = list(iter.max = 1000, eval.max = 1000), profile = TRUE), data =  IC_comp_data)`

Thanks!

[IC_comp_data.txt](https://github.com/glmmTMB/glmmTMB/files/2464933/IC_comp_data.txt)

**Comment by bbolker** [comment]
**tl;dr** 

(1) this is an ambitious modeling exercise - you have only 26 'successes' out of 321 data points, so fitting a model with 2 fixed effects and several random effects is fairly dicey to begin with. Leaving tou `profile=TRUE` does get past the specific hurdle you mention above.  After boiling the model down enough to fit it, we still get results that look questionable.
(2) there is a specific problem with NAs popping up unexpectedly when `NA=TRUE` which we could try to fix.

## Modeling issues

Full code below.  No matter what we do, we get:

- very large-magnitude parameter estimates, which are usually a sign of complete separation in binomial-response models
- a subset of the conditional modes (for both Year and Pair_ID) are extreme, presumably because there are a bunch of years and a bunch of pairs for which there are no successes

I don't know quite what to recommend for this analysis; you could try to get away with a GLM fit, but I'm not sure that's reliable either (since there does actually appear to be a fair amount of variation among years & pairs).

## Step-by-step:

I don't actually get convergence warnings with your first model, although I do get some very small variance components (`BandM` and `Year` have standard devs of approximately 1, while `Site` ~ 1e-4 and `BandF` ~ 1e-7. I can replicate your error with the next model; the proximal problem is [here](https://github.com/glmmTMB/glmmTMB/blob/master/glmmTMB/R/glmmTMB.R#L847), where some attempts to take some Newton-Raphson steps (with a fixed Hessian?) apparently fail. (I don't really know what the `profile` option is doing; it was introduced [here](https://github.com/glmmTMB/glmmTMB/commit/f10030996460e9fb02f7e07fba66c811d02157a5). @kaskr?)

It would obviously be possible to add some code to check for `NA` values and bail out of the Newton steps, but I don't know if that would be papering over cracks?

My dumb advice would be to leave out the `profile=TRUE` (docs say this is recommended to improve fitting "when a model has many fixed effects" ...) When I do this, I get non-positive-definite Hessian warnings and the following results:

```
VarCorr(mod_TMB_Qlty2P)

Conditional model:
 Groups Name        Std.Dev. 
 PairID (Intercept) 105.03017
 Site   (Intercept)   0.11166
 Year   (Intercept)  57.99103
> coef(summary(mod_TMB_Qlty2) )
$cond
              Estimate Std. Error   z value     Pr(>|z|)
(Intercept) -54.818433  11.098789 -4.939136 7.846929e-07
stdQlty       8.727464   4.137215  2.109502 3.490125e-02
stdLaying    10.400946   4.569820  2.276008 2.284555e-02
```

## Full code

```
dd <- read.table("IC_comp_data.txt",header=TRUE)
library(glmmTMB)
library(ggplot2); theme_set(theme_bw())
library(ggalt) ## for geom_encircle

table(dd$ProbDiv)  ## n=26;

## 2D plot (biv effects of Qlty + laying) - no obvious problems
ggplot(dd,aes(stdQlty,stdLaying))+geom_point(aes(colour=factor(ProbDiv)))+
    geom_encircle(aes(group=Year),colour="gray")+
    scale_colour_manual(values=c("black","red"))

## univariate effects of Qlty
ggplot(dd,aes(stdQlty,ProbDiv))+stat_sum(alpha=0.5)+
    geom_smooth()+
    geom_smooth(method="glm",
                method.args=list(family=binomial),
                colour="red")

## full model
mod_TMB_Qlty <- glmmTMB(ProbDiv ~ stdQlty + stdLaying +
                            (1|Year) + (1|Site) + (1|BandM) + (1|BandF),
                        family = "binomial",
                        control = glmmTMBControl(optCtrl = list(iter.max = 1000,
                                        eval.max = 1000), profile = TRUE),
                        data = dd)

coef(summary(mod_TMB_Qlty))
VarCorr(mod_TMB_Qlty)  ## BandF, Site ~ 0

## Wald effects probably not reliable: profile confidence intervals suggest we can't
##  actually see a damn thing ...
confint(mod_TMB_Qlty2P,method="uniroot", parm=2:3)

## reduced models
try(mod_TMB_Qlty2 <- update(mod_TMB_Qlty,
                        . ~ . - (1|BandM) - (1|BandF) + (1|PairID)))

## turn off profile
mod_TMB_Qlty2P <- update(mod_TMB_Qlty,
                         . ~ . - (1|BandM) - (1|BandF) + (1|PairID),
               control = glmmTMBControl(optCtrl = list(iter.max = 1000,
                                  eval.max = 1000), profile = FALSE))
                         

coef(summary(mod_TMB_Qlty2P))
VarCorr(mod_TMB_Qlty2P)

## without Site
mod_TMB_Qlty3 <- update(mod_TMB_Qlty2P, . ~ . - (1|Site))
coef(summary(mod_TMB_Qlty3))
VarCorr(mod_TMB_Qlty3)

## regular GLM does fit, without any apparent pathology (so complete sep. is *not*
## in the fixed-effects part of the model)
summary(glm(ProbDiv ~ stdQlty+stdLaying, family=binomial, data=dd))

## plot ranefs -- looks bad!
rr <- ranef(mod_TMB_Qlty3)
pp <- lme4:::dotplot.ranef.mer(rr$cond)
gridExtra::grid.arrange(pp[[1]],pp[[2]])

## Laplace approximation is known to be bad in the binary, small-cluster case! 
##  I wanted to try GLMMadaptive - thought it would handle multiple independent RE terms in a model, ## but AFAICT it doesn't (it does handle vector-valued RE terms, which lme4 won't ...) - 
## so this model could also be done in lme4::glmer, with (say) nAGQ=15, to turn on 
## Gauss-Hermite quadrature [15 q points is GLMMadaptive's default]

library(GLMMadaptive)
mm <- mixed_model(fixed = ProbDiv ~ stdQlty + stdLaying ,
                  random = ~ 1 | Year,
     family = binomial(),
     data = dd)

plot(sort(ranef(mm)))
## answer: better, but still looks dicey

## started trying a Bayesian analysis with regularizing priors ...

library(brms)
form <- ProbDiv ~ stdQlty + stdLaying +(1|Year) + (1|Site) + (1|PairID)
prior <- get_prior(form, family=bernoulli, data=dd)
## student-t priors: (df, mean, sd)
prior$prior[1] <- "student_t(3, 0, 5)"  ## fixed effects
prior$prior[4] <- "student_t(3, 0, 5)"  ## intercept (default is sd=10)
prior$prior[5] <- "student_t(3, 0, 5)"  ## RE sds (default is sd=10)
b1 <- brm(form, prior=prior, 
          family = bernoulli,
          data = dd)

```

**Comment by tmerkling** [comment]
Thanks @bbolker for your help!

It looks like I don't really have the choice but to simplify the RE structure (and even that won't solve all the problems if I understand correctly). So maybe just focus on PairID and Year on the basis that there is more variation substantial variation for those but much less for Site ?

When I got those warning and error messages, I also tried Bayesian analyses with brms. I got some issues with my priors too uninformative and was advised to maybe use half-normal priors or half-student t (as I think you did), but not half-Cauchy (to avoid having a fat tails). Would you say that making use of the flexibility of the Bayesian framework is the 'least bad' way of analysing those data?
Would it still be worth comparing ranking of the different models and relative magnitude of the different estimates with GLMs?

**Comment by bbolker** [comment]
I would say that Bayes is probably the "least bad" approach, but it's not because of the flexibility; it's because of the ability to *regularize*, i.e. adding information that prevents the model from reaching silly conclusions based on a a small data set.  You really have to confront the fact that (despite whatever heroic efforts it may have taken to collect this data set) you don't have enough data to estimate variability among groups in addition to the fixed effects.  There's a catch-22 here: if you're supposed to pay attention to group-level variability because it might affect your conclusions about your fixed effects, what do you do when your data set isn't big enough? 

The GLMs will tell you *something* about the magnitude of effects, but they may underestimate the degree of uncertainty.

**Comment by bbolker** [comment]
Leaving this open for now because we do have a technical issue, which is that when we get an NA in the course of profiling, the Newton-Raphson loop should fail with a more informative error. Bumping this for @kaskr: it would be fairly easy to add logic [here](https://github.com/glmmTMB/glmmTMB/blob/master/glmmTMB/R/glmmTMB.R#L861) that would bail out of the profiling step and behave as though the Hessian had been non-pos-def in the first place if a non-finite value was detected in the loop, but I don't know if papering over cracks in that way is the appropriate strategy ... should we try to dig in and analyze what went wrong in this particular case?  (It's not really surprising that Newton-Raphson could diverge if the model fit is relatively unstable in the first place ...)

> the proximal problem is [here](https://github.com/glmmTMB/glmmTMB/blob/master/glmmTMB/R/glmmTMB.R#L847), where some attempts to take some Newton-Raphson steps (with a fixed Hessian?) apparently fail. (I don't really know what the `profile` option is doing; it was introduced [here](https://github.com/glmmTMB/glmmTMB/commit/f10030996460e9fb02f7e07fba66c811d02157a5))

**Comment by kaskr** [comment]
@bbolker A comment on what the profile option is doing: The main idea is partially described [here](https://kaskr.github.io/adcomp/_book/Appendix.html). The way it is being used in glmmTMB is to provide an approximate solution that can be used as starting guess for the usual glmmTMB objective function. The profile option of glmmTMB has the following properties (TODO @kaskr: document it).
* For Gaussian models `profile=TRUE` and `profile=FALSE` give the exact same results.
* For other models the `profile=TRUE` and `profile=FALSE` should give exact same *estimates* but *uncertainties* may differ slightly (because they are based on the profiled hessian in the profile case).
* The profile option is meant to handle cases with *many fixed effects* (say more than 10) and plenty of data such that the Gaussian approximation of the posterior is reasonable (the present case doesn't appear to belong in this category !).

It may be worth digging into the Newton steps for this case to understand what is going on and possibly robustify (will do but not now).
